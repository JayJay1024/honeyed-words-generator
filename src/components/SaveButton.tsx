import React, { useState } from 'react';
import styled from 'styled-components';
import html2canvas from 'html2canvas';
import { saveAs } from 'file-saver';
import ImageDownload from '../assets/imgs/download.svg';
import StyledButton from './StyledButton';

const ua = navigator.userAgent;
const isiOSwebview = /(iPhone|iPod|iPad).*AppleWebKit(?!.*Safari)/i.test(ua);
const isWebview = ua.toLowerCase().indexOf('micromessenger') > -1 || isiOSwebview;

const Button = styled<any>(StyledButton)`
    position: fixed;
    left: 0.5rem;
    bottom: 0.5rem;

    background-image: url(${ImageDownload});
`;

const SaveButton: React.FunctionComponent<{visible: boolean}> = ({visible}): JSX.Element => {
    const [generating, setGenerating] = useState(false);

    const generateImage = async (ele: Element, isWebview: boolean = false) => {
        setGenerating(true);

        html2canvas(ele as HTMLElement, {
            onclone: document => {
                let tmp: HTMLElement = document.querySelector('#HONEYED_WORDS_CARD');
                tmp.classList.add('starting');
                tmp.style.transform = 'none';
                tmp.style.boxShadow = 'none';
            },
            scale: window.devicePixelRatio * (isWebview ? 2 : 1)
        })
        .then((canvas) => {
            if (isWebview) {
                let img = document.createElement('img');

                canvas.toBlob(blob => {
                    const {
                        URL: { createObjectURL }
                    } = window;
                    img.src = createObjectURL(blob);
                    img.classList.add('downloadImg');
                });
                ele.classList.add('img');
                ele.appendChild(img);
                setGenerating(false);
            } else {
                canvas.toBlob(blob => {
                    saveAs(blob, `hw-${new Date().getTime()}.png`);
                    setGenerating(false);
                }, 'image/png');
            }
            ele.classList.remove('starting');
        });
    };

    const handleDownload = async () => {
        if (isWebview) {
            alert('请在浏览器打开！');
        }
        let ele = document.querySelector('#HONEYED_WORDS_CARD');
        await generateImage(ele);
    };

    return (
        <Button
            className={visible ? 'visible' : 'hidden'}
            disabled={generating}
            onClick={handleDownload}
        ></Button>
    );
};

export default SaveButton;
