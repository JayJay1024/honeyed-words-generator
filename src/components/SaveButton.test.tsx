import React from 'react';
import { create, act, ReactTestRenderer } from 'react-test-renderer';

import SaveButton from './SaveButton';

describe('component RefreshButton', () => {
    let testRenderer: ReactTestRenderer = null;

    beforeEach(() => {
        act(() => {
            testRenderer = create(<SaveButton visible={false} />);
        });
    });

    afterEach(() => {
        if (testRenderer) {
            testRenderer.unmount();
            testRenderer = null;
        }
    });

    test('RefreshButton should match snapshot', () => {
        let tree = testRenderer.toJSON();
        expect(tree).toMatchSnapshot();

        act(() => {
            testRenderer.update(<SaveButton visible={true} />);
        });
        tree = testRenderer.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('there should be a button', () => {
        const testInstance = testRenderer.root;
        expect(testInstance.findAllByType("button").length).toEqual(1);
    });
});
