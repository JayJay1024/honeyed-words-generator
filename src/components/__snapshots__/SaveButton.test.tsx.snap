// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component RefreshButton RefreshButton should match snapshot 1`] = `
<button
  className="sc-AxjAm sc-AxirZ OFvGj hidden"
  disabled={false}
  onClick={[Function]}
/>
`;

exports[`component RefreshButton RefreshButton should match snapshot 2`] = `
<button
  className="sc-AxjAm sc-AxirZ OFvGj visible"
  disabled={false}
  onClick={[Function]}
/>
`;
